codec/src/main/java/io/netty/handler/codec/compression/ZstdEncoder.java

```diff
@@ -0,0 +1,194 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import com.github.luben.zstd.Zstd;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.EncoderException;
+import io.netty.handler.codec.MessageToByteEncoder;
+import io.netty.util.internal.ObjectUtil;
+import java.nio.ByteBuffer;
+
+import static io.netty.handler.codec.compression.ZstdConstants.*;
+
+/**
+ *  Compresses a {@link ByteBuf} using the Zstandard algorithm.
+ *  See <a href="https://facebook.github.io/zstd">Zstandard</a>.
+ */
+public final class ZstdEncoder extends MessageToByteEncoder<ByteBuf> {
+
+    private final int blockSize;
+    private final int compressionLevel;
+    private final int maxEncodeSize;
+    private ByteBuf buffer;
+
+    /**
+     * Creates a new Zstd encoder.
+     *
+     * Please note that if you use the default constructor, the default BLOCK_SIZE and MAX_BLOCK_SIZE
+     * will be used. If you want to specify BLOCK_SIZE and MAX_BLOCK_SIZE yourself,
+     * please use {@link ZstdEncoder(int,int)} constructor
+     */
+    public ZstdEncoder() {
+        this(DEFAULT_COMPRESSION_LEVEL, DEFAULT_BLOCK_SIZE, MAX_BLOCK_SIZE);
+    }
+
+    /**
+     * Creates a new Zstd encoder.
+     *  @param  compressionLevel
+     *            specifies the level of the compression
+     */
+    public ZstdEncoder(int compressionLevel) {
+        this(compressionLevel, DEFAULT_BLOCK_SIZE, MAX_BLOCK_SIZE);
+    }
+
+    /**
+     * Creates a new Zstd encoder.
+     *  @param  blockSize
+     *            is used to calculate the compressionLevel
+     *  @param  maxEncodeSize
+     *            specifies the size of the largest compressed object
+     */
+    public ZstdEncoder(int blockSize, int maxEncodeSize) {
+        this(DEFAULT_COMPRESSION_LEVEL, blockSize, maxEncodeSize);
+    }
+
+    /**
+     * @param  blockSize
+     *           is used to calculate the compressionLevel
+     * @param  maxEncodeSize
+     *           specifies the size of the largest compressed object
+     * @param  compressionLevel
+     *           specifies the level of the compression
+     */
+    private ZstdEncoder(int compressionLevel, int blockSize, int maxEncodeSize) {
+        super(true);
+        this.compressionLevel = ObjectUtil.checkPositive(compressionLevel, "compressionLevel");
+        this.blockSize = blockSize;
```

Comment:
Should we verify this is positive as well ?

Reply:
I've done it